roomserver/storage/shared/storage.go

```diff
@@ -1589,6 +1591,29 @@ func (d *Database) UpgradeRoom(ctx context.Context, oldRoomID, newRoomID, eventS
 	})
 }
 
+// InsertUserRoomKey inserts a new user room key for the given user and room.
+// Returns an error if a database error occurred, also if the primary constraint was violated.
+func (d *Database) InsertUserRoomKey(ctx context.Context, userNID types.EventStateKeyNID, roomNID types.RoomNID, key ed25519.PrivateKey) error {
```

Comment:
This API shape isn't the most helpful and is racey.

The main scenario we are designing for is "I want to join a pseudo ID room, please ensure there is a key for me". If you concurrently call `InsertUserRoomKey`, the slower call will fail with an error, when in reality you just want to return the key that was previously inserted. I would probably change this to return the private key to use and make use of `ON CONFLICT ... RETURNING` like we do elsewhere in Dendrite.

Reply:
In theory it shouldn't race when called through `/join`, since there can only be one join in progress.